drivers/gpio/gpio_gecko.c

```diff
@@ -91,18 +93,31 @@ static int gpio_gecko_configure(struct device *dev,
 		return -ENOTSUP;
 	}
 
-	if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
-		if ((flags & GPIO_PUD_MASK) == GPIO_PUD_PULL_UP) {
+	if (flags & GPIO_OUTPUT) {
```

Comment:
When I was looking at the flags, I wondered what the code would do when both GPIO_OUTPUT and GPIO_INPUT were set.

This driver respects output before input, but I don't see it documented that OUTPUT always has higher priority if both are set. Is that your intent?

I think it should be documented if so. Though my preference would be slightly for rejecting this as an invalid combination -- perhaps a common helper function, `gpio_flags_are_valid(u32_t flags)`, could be added for this purpose?

Reply:
There are several devices where it's perfectly valid and very useful to have a particular GPIO configured for input and output simultaneously.  The API should allow that, on those devices.

Reply:
Fair enough. What are the intended semantics if the user requests both but the driver only supports one?

Reply:
Dunno; `-ENOTSUP` would seem to make sense, though `-EINVAL` is used here for a combination that would be acceptable on other platforms (`GPIO_INT` with `GPIO_OUTPUT`should be fine if `GPIO_INPUT` is also set).

In this case the comments suggest that bidirectional is handled in `GPIO_OUTPUT` case, though I don't see how that works and it's not how I would expect it to be expressed for Nordic or SX1509B.  In those input and output are simply independent functions.

The API documentation needs to be reviewed and updated to describe exactly what is and is not allowed, and what behavior the different combinations have to provide in order to meet the API's promises.  `GPIO_DEBOUNCE` is one I noticed that still has about the same chance for misinterpretation as `GPIO_POL_INV` did in the current API.

Reply:
`-ENOTSUP` sounds like a good plan to me for unsupported (but valid) combinations of flags.

Reply:
This bit of code was written specifically for the gecko GPIO module. In case any of the output modes are enabled the input will also be implicitly enabled. This is actually a fairly common approach among SoC vendors but surely not the only one. Application code that wants to read from the input must pass `GPIO_INPUT` flag to the configure function. In case of the gecko platform if it passes `GPIO_OUTPUT` only the input direction will still work. A portable application code shouldn't rely on this 'feature'. The gecko GPIO driver doesn't perform extra checks, e.g. the implementation of `gpio_pin_get` function doesn't check if the input direction was explicitly configured as this would increase function's code size, make it slower.

As for the `GPIO_OUTPUT`, `GPIO_INPUT` flag usage. If an application wants to write to the pin it should pass `GPIO_OUTPUT` flag to the configure function, if it intends to read from the pin it should pass `GPIO_INPUT` flag. If it intends to write and read it should pass both flags. If no flag is passed the pin will be disabled.

In general, if the function is not able to provide features requested by the configuration flags it should return `-ENOTSUP`.

I agree though we should better document error code / assert usage.

Reply:
> I agree though we should better document error code / assert usage.

That would be great if you don't mind. I really like your proposal of adding asserts into gpio.h itself in addition to docs. :+1: